src/compiler/moduleSpecifiers.ts

```diff
@@ -147,7 +147,9 @@ function getPreferences(
                 return [ModuleSpecifierEnding.JsExtension];
             }
             if (getEmitModuleResolutionKind(compilerOptions) === ModuleResolutionKind.Classic) {
-                return [ModuleSpecifierEnding.Index, ModuleSpecifierEnding.JsExtension];
+                return preferredEnding === ModuleSpecifierEnding.JsExtension
+                    ? [ModuleSpecifierEnding.JsExtension, ModuleSpecifierEnding.Index]
+                    : [ModuleSpecifierEnding.Index, ModuleSpecifierEnding.JsExtension];

```

Comment:
This change alone fixes the issue reported by @jespertheend in #52167, which was different from the OP’s issue. When `--moduleResolution` is `classic` (the unfortunate default of `--module esnext`, we simply weren’t respecting the inferred preference. `classic` resolution doesn’t support dropping `/index.js` suffixes, which is why it gets its own if block here that never returns `ModuleSpecifierEnding.Minimal`.